### ISSUE TITLE
Auto connect/disconnect to VPN based on app(s)

### ISSUE THREAD

**Issue by pplucky** [zaneschepke/wgtunnel#44]
Hello.

First of all, thanks a lot for your great app. Clearly the AndroidTV flavor of Wireguard is lacking some needed functionalities.

I know that like in Wireguard original app (worst case scenario by manually editing tbe .conf file before importing), it is possible to define specific apps to which a specific tunnel is applicable or not. 

Still then, this implies one of two things:
- having tunnel always-on and allow system to apply tunnel only to specified apps:
- manually turning on/off the tunnel, when required (before using the app(s).

Having that into consideration, I wonder if it would be a possibility to have the tunnel activated/deactivated every time the specified apps are opened/closed. 

I know this can eventually be done using adb, but I wonder if implementing such a functionality would be feasible within the service and if you could consider implementing it.

Thanks in advance.

**Comment by zaneschepke** [comment]
Hello! Thank you for using the app!

Can you provide me with an example use case to better understand this request?

**Comment by pplucky** [comment]
Hi there.

Yes, of course, here are 2 examples of use cases:
- To use specific apps, I need to be connected to a tunnel at work, so I have to activate the work tunnel before opening the apps and deactivate it after closing the apps;
- When I'm on vacation, I use specific apps to watch movies from my home library, so I have to activate the home tunnel before opening Kodi (for instance) and deactivate it after closing it.

My idea would be for the second use case (taking Kodi as the example app):
- Manually open Kodi app;
- When Kodi app was opened, WG Tunnel service would notice it and activate the "Home tunnel";
- Who Kodi app is closed (or at least not the foreground app anymore), WG Tunnel service would notice it and de-activate the "Home tunnel".

That was basically my idea for the use cases (for sure there are others).

Thanks in advance, once more.

**Comment by zaneschepke** [comment]
Ah! Okay, I understand the use case now. 

This can be accomplished using Macrodroid + Shortcut Maker (Macrodroid can trigger off of application launches) or other similar automation apps integrated with WG Tunnel application shortcuts. I have tested this integration, and I know it works. 

However, the app currently falls short of the multiple tunnel use case. The app only has one on and one off shortcut for the primary tunnel (a static shortcut).  It would need dynamic shortcuts (or a different way of integrating) working with these automation tools so you can toggle specific tunnels based on the launched application. I'll look into this. 

This would be the easiest solution to maintain flexibility with the triggers while still accomplishing your use case. 

Thoughts?

**Comment by zaneschepke** [comment]
In version 3.1.6, you can now toggle tunnels by name via intent. 

This is also relevant for #19. 

Details: 
Target: Activity
Action: (START or STOP)
Package: com.zaneschepke.wireguardautotunnel
Class: com.zaneschepke.wireguardautotunnel.service.shortcut.ShortcutsActivity

There are two required "extras" (these are key-value pairs):

key: className
value: WireGuardTunnelService

key: tunnelName
value: (name of your tunnel)

Right now, these are undocumented. I will work on a way to document them. 

I have tested these in MacroDroid, and it works well.

MacroDroid has a simple setup for app launch triggers to fit your use case. 

If you run into any problems, we can reopen this issue.

**Comment by Neo-Oli** [comment]
To anyone coming here looking for instructions. Here's what I learned. 

How to send that intent using `am`, which can be used with adb or termux

Start a tunnel named wg0:
```bash
    am start \
        -n com.zaneschepke.wireguardautotunnel/com.zaneschepke.wireguardautotunnel.service.shortcut.ShortcutsActivity \
        -a START \
        --es className WireGuardTunnelService \
        --es tunnelName "wg0"
```
Stop a tunnel named wg0:
```bash
    am start \
        -n com.zaneschepke.wireguardautotunnel/com.zaneschepke.wireguardautotunnel.service.shortcut.ShortcutsActivity \
        -a STOP \
        --es className WireGuardTunnelService \
        --es tunnelName "wg0"
```

If you are migrating from the other wireguard app (`com.wireguard.android`), I also noticed this difference:

Calling the `START` action on a tunnel that is already up, it will turn it off and back on, as opposed to ignoring the intent. This can cause short disconnects. So if you are turning on the tunnel in some kind of script, make sure the tunnel is not already up.

If a tunnel exists can be checked with 
```bash
if ip link show tun0 > /dev/null 2>&1; then 
   echo "Connected to a tunnel"
fi
```

So far I haven't found a way to find out which tunnel is connected from a script.

**Comment by zaneschepke** [comment]
> To anyone coming here looking for instructions. Here's what I learned.
> 
> How to send that intent using `am`, which can be used with adb or termux
> 
> Start a tunnel named wg0:
> 
> ```shell
>     am start \
>         -n com.zaneschepke.wireguardautotunnel/com.zaneschepke.wireguardautotunnel.service.shortcut.ShortcutsActivity \
>         -a START \
>         --es className WireGuardTunnelService \
>         --es tunnelName "wg0"
> ```
> 
> Stop a tunnel named wg0:
> 
> ```shell
>     am start \
>         -n com.zaneschepke.wireguardautotunnel/com.zaneschepke.wireguardautotunnel.service.shortcut.ShortcutsActivity \
>         -a STOP \
>         --es className WireGuardTunnelService \
>         --es tunnelName "wg0"
> ```
> 
> If you are migrating from the other wireguard app (`com.wireguard.android`), I also noticed this difference:
> 
> Calling the `START` action on a tunnel that is already up, it will turn it off and back on, as opposed to ignoring the intent. This can cause short disconnects. So if you are turning on the tunnel in some kind of script, make sure the tunnel is not already up.
> 
> If a tunnel exists can be checked with
> 
> ```shell
> if ip link show tun0 > /dev/null 2>&1; then 
>    echo "Connected to a tunnel"
> fi
> ```
> 
> So far I haven't found a way to find out which tunnel is connected from a script.

Thanks for documenting this!

If it is helpful, I can change this to ignore the intent if the tunnel is already up so you don't need to know the status.

**Comment by dkoppenh** [comment]
I agree it's more intuitive to ignore the intent if the tunnel is already up.

**Comment by BlackWolfWoof** [comment]
I wasn't able to control the app via Intents from Tasker sadly.
This is my current setup:
https://github.com/user-attachments/assets/d7c81245-f609-4196-b98d-6d38ae03f29e

**Comment by BlackWolfWoof** [comment]
3.7.1 is the app version, and i got it from F-Droid

**Comment by zaneschepke** [comment]
> 3.7.1 is the app version, and i got it from F-Droid

Likely, this is fixed in 3.7.2 which should be on F-Droid soon. if it is not, please open a new issue for this and I'll investigate. Thank you!

**Comment by BlackWolfWoof** [comment]
> > 3.7.1 is the app version, and i got it from F-Droid
> 
> Likely, this is fixed in 3.7.2 which should be on F-Droid soon. if it is not, please open a new issue for this and I'll investigate. Thank you!

I got it to work. I checked out the XML and found that it is called com.zaneschepke.wireguardautotunnel.**core**.shortcut.ShortcutsActivity, which was missing before.
I. Addition, since i never started a tunnel, it didn't ask me for the VPN permission prompt and therefore didn't work on the first run. After manually toggling it once in the app though it worked flawlessly.


```bash
am start \
        -n com.zaneschepke.wireguardautotunnel/com.zaneschepke.wireguardautotunnel.core.shortcut.ShortcutsActivity \
        -a START \
        --es className WireGuardTunnelService \                                    --es tunnelName "Wagon"
```

```bash
am start \
        -n com.zaneschepke.wireguardautotunnel/com.zaneschepke.wireguardautotunnel.core.shortcut.ShortcutsActivity \
        -a STOP \
        --es className WireGuardTunnelService \                                    --es tunnelName "Wagon"
```

This can be translated to Tasker
![Screenshot_2025-03-25-06-38-36-65.jpg](https://github.com/user-attachments/assets/12c01d20-8aef-4b0a-9224-a07afc57bfde)

**Comment by BlackWolfWoof** [comment]
The app got updated, i am now on 3.8.1 and what i did above no longer works. What got changed?
